### PR TITLE
557 domain function mapping files

### PR DESCRIPTION
**Pull request by Jday7879**
# Summary
Functions have been added to derive the SIC - Domain mapping functions on sharepoint
Functions to produce the output for selective editing have been produced. 
Minor changes to selective editing contributing method to join threshold values

# Checklists

This pull request meets the following requirements:

- [ ] installable with all dependencies recorded
- [x] runs without error
- [x] follows PEP8 and project specific conventions
- [ ] appropriate use of comments, for example no descriptive comments
- [ ] functions documented using Numpy style docstings
- [ ] assumptions and decisions log considered and updated if appropriate
- [ ] unit tests have been updated to cover essential functionality for a reasonable range of inputs and conditions
- [ ] other forms of testing such as end-to-end and user-interface testing have been considered and updated as required
- [ ] tests suite passes (locally as a minimum)
- [ ] peer reviewed with review recorded

If you feel some of these conditions do not apply for this pull request, please
add a comment to explain why.

### CODE REVIEW

File: mbs_results/outputs/selective_editing_question_output.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import pandas as pd
+
+from mbs_results.merge_domain import merge_domain
+from mbs_results.unsorted.selective_editing import (
+    calculate_predicted_value,
+    create_standardising_factor,
+)
+
+
+def create_selective_editing_question_output(
```
Comment (AntonZogk) [inline mbs_results/outputs/selective_editing_question_output.py:12]: type hints for some args missing

File: mbs_results/outputs/selective_editing_question_output.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    creates and saves the selective editing question output to the output_path folder
+    saves it as a csv with the same version tag
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        input data, this uses the output produced by the winsorisation functions
+    output_path : str
+        path to save the output csv. Output file path must be the sharepoint as this
+        path is also used to load needed mapping files (could change to individual arg?)
```
Comment (AntonZogk) [inline mbs_results/outputs/selective_editing_question_output.py:39]: path is dependent to the current folder structure in sharepoint ? What is going to happen if it changes ?

File: mbs_results/outputs/selective_editing_question_output.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    version : str
+        version of the monthly-business-survey-results package, used to tag the output
+        with the correct version of the release
+    previous_period : int
+        previous period to take the weights for estimation of standardising factor in
+        the format yyyymm
+    """
+    sic_domain_mapping = pd.read_csv(
```
Comment (AntonZogk) [inline mbs_results/outputs/selective_editing_question_output.py:47]: Could do a chain with pipe perhaps (optional)

File: mbs_results/outputs/selective_editing_question_output.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        predicted_value=predicted_value,
+        imputation_marker="imputation_flags_adjusted_value",
+        a_weight=a_weight,
+        o_weight=o_weight,
+        g_weight=g_weight,
+        auxiliary_value=aux,
+        period_selected=period_selected,
+    )
+
+    standardising_factor["survey_code"] = "009"
```
Comment (AntonZogk) [inline mbs_results/outputs/selective_editing_question_output.py:78]: Why is this needed?

File: mbs_results/outputs/selective_editing_question_output.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    validation_checks_selective_editing(output)
+    formatted_date = datetime.today().strftime("%Y-%m-%d")
+    output_file_name = (
+        f"sopp_mbs_{formatted_date}_selective_editing"
+        + f"_contributor_{previous_period}_{version}.csv"
+    )
+    output.to_csv(
+        output_path + "selective_editing_outputs/" + output_file_name,
+        index=False,
+    )
```
Comment (AntonZogk) [inline mbs_results/outputs/selective_editing_question_output.py:153]: This is how the user will call the outputs function right ?  We could remove the main bit and add it as an example how to use it ?

File: mbs_results/selective_editing_contributer_output.py [hunk 6]
```diff
@@ -34,22 +38,61 @@ def get_selective_editing_contributer_output(
...
+        threshold_mapping,
+        left_on=["form_type", "domain"],
+        right_on=["form", "domain"],
+        how="left",
+    ).drop(columns=["form", "form_type"])
+
     selective_editing_contributer_output = selective_editing_contributer_output.rename(
         columns={"reference": "ruref", "domain": "domain_group"}
     )
+    selective_editing_contributer_output["survey_code"] = "009"
```
Comment (AntonZogk) [inline mbs_results/selective_editing_contributer_output.py:63]: Same as above why  add a new column with values 009 ?

File: mbs_results/selective_editing_contributer_output.py [hunk 7]
```diff
@@ -34,22 +38,61 @@ def get_selective_editing_contributer_output(
...
+
+    return selective_editing_contributer_output.loc[
+        selective_editing_contributer_output["period"] == period_selected
+    ]
 
-    return selective_editing_contributer_output
+
+if __name__ == "__main__":
```
Comment (AntonZogk) [inline mbs_results/selective_editing_contributer_output.py:70]: Same comment as above for this main

File: mbs_results/unsorted/selective_editing.py [hunk 8]
```diff
@@ -91,9 +91,8 @@ def create_standardising_factor(
         each reference.
 
     """
-    previous_df = dataframe[(dataframe[period] == previous_period)]
-    previous_df = previous_df[previous_df[question_code].isin([40, 49])]
-
+    previous_df = dataframe[(dataframe[period] == period_selected)]
+    previous_df = previous_df[previous_df[question_no].isin([40, 49])]
```
Comment (AntonZogk) [inline mbs_results/unsorted/selective_editing.py:95]: Is it going to be always 40 49 ? If so could add this on top as a constant to be clear it won't be changed

File: mbs_results/utilities/derive_mapping_files.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+import pandas as pd
+
+# Run this file to re derive the sic to domain mapping file from sharepoint.
+
+
+def derive_sic_domain_mapping(
```
Comment (AntonZogk) [inline mbs_results/utilities/derive_mapping_files.py:6]: Docstrings missing

File: mbs_results/utilities/derive_mapping_files.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    # Outputting matched and unmatched files to sharepoint
+    sic_domain_map.to_csv(output_path + "sic_domain_mapping.csv", index=False)
+    unmatched_sic.to_csv(output_path + "sic_domain_unmatched_sic.csv", index=False)
+    unmatched_domain.to_csv(
+        output_path + "sic_domain_unmatched_domain.csv", index=False
+    )
+
+
+if __name__ == "__main__":
```
Comment (AntonZogk) [inline mbs_results/utilities/derive_mapping_files.py:49]: Same as other mains, could add as an example in docs?

**Review by AntonZogk — CHANGES_REQUESTED** [review]
All looking good. Some minor changes , just i think we need to removed the if __name__ == "__main__" and the code in it, but perhaps we could add that bit as an example in the docs. 

If the script has  if __name__ == "__main__"  normally i am expecting a runnable file which can be executed from the cmd.  Would love to hear your thoughts on this :)

File: mbs_results/outputs/selective_editing_question_output.py (hunk 2 above)
Comment (Jday7879) [reply mbs_results/outputs/selective_editing_question_output.py:39]: Realised that I refactored code and it no longer saves the output, instead it returns it. Renaming this variable to be mapping file path instead to avoid this issue

**Comment by Jday7879** [comment]
Made the changes you suggested @AntonZogk. I will move the script to create the outputs to the testing branch so they can be produced easily next week following 0.0.3 release. I'll create a branch for that now. Ready for this to be reviewed again :)

**Comment by Jday7879** [comment]
Forgot to push before re requesting my bad!

**Review by AntonZogk — APPROVED** [review]
Thanks for updating @Jday7879 , all looking good happy to be merged :)
